Let's set up an example scenario as perspective for our regression exercises using the Zillow dataset.

As a Codeup data science graduate, you want to show off your skills to the Zillow data science team in hopes of getting an interview for a position you saw pop up on LinkedIn. You thought it might look impressive to build an end-to-end project in which you use some of their Kaggle data to predict property values using some of their available features; who knows, you might even do some feature engineering to blow them away. Your goal is to predict the values of single unit properties using the observations from 2017.

In these exercises, you will complete the first step toward the above goal: acquire and prepare the necessary Zillow data from the zillow database in the Codeup database server.

In [2]:
# import necessary libraries
import pandas as pd
import numpy as np
import os
from tqdm import tqdm

# import personal env file
import env

In [3]:
# Use personally developed function to create a wrangle.py file
# WARNING: FUNCTION WILL OVERWRITE ANY 'wrangle.py' FILE THAT EXISTS
env.deploy_wrangle()

In [4]:
# import newly built wrangle file
import wrangle as w

In [5]:
# create write_contents
# write_contents will hold functions I'm going to add to wrangle.py
write_contents = []

1. [x] Acquire `bedroomcnt`, `bathroomcnt`, `calculatedfinishedsquarefeet`, `taxvaluedollarcnt`, `yearbuilt`, `taxamount`, and `fips` from the `zillow` database for all 'Single Family Residential' properties.

- [x] Use MySQL Workbench to isolate the necessary dataset

In [8]:
# build query
query = """
select bedroomcnt, bathroomcnt, calculatedfinishedsquarefeet, taxvaluedollarcnt, yearbuilt, taxamount, fips
from propertylandusetype
	join properties_2017
		using (propertylandusetypeid)
WHERE propertylandusedesc = ("Single Family Residential")
"""

In [9]:
# use env file to get database url
url = env.get_db_url('zillow')

In [10]:
# use pandas SQL query command to get the DataFrame
df = pd.read_sql(query,url)

In [11]:
df.head()

bedroomcnt  bathroomcnt  calculatedfinishedsquarefeet  taxvaluedollarcnt  \
0         0.0          0.0                           NaN            27516.0   
1         0.0          0.0                           NaN               10.0   
2         0.0          0.0                           NaN               10.0   
3         0.0          0.0                           NaN             2108.0   
4         4.0          2.0                        3633.0           296425.0   

   yearbuilt  taxamount    fips  
0        NaN        NaN  6037.0  
1        NaN        NaN  6037.0  
2        NaN        NaN  6037.0  
3        NaN     174.21  6037.0  
4     2005.0    6941.39  6037.0

In [12]:
w.check_file_exists('zillow.csv',query,url)

Reading from file...


bedroomcnt  bathroomcnt  calculatedfinishedsquarefeet  \
0               0.0          0.0                           NaN   
1               0.0          0.0                           NaN   
2               0.0          0.0                           NaN   
3               0.0          0.0                           NaN   
4               4.0          2.0                        3633.0   
...             ...          ...                           ...   
2152858         4.0          3.0                        2262.0   
2152859         4.0          4.5                        3127.0   
2152860         0.0          0.0                           NaN   
2152861         3.0          2.5                        1974.0   
2152862         4.0          4.0                        2110.0   

         taxvaluedollarcnt  yearbuilt  taxamount    fips  
0                  27516.0        NaN        NaN  6037.0  
1                     10.0        NaN        NaN  6037.0  
2                     10.0        NaN        NaN  6037.0  
3                   2108.0        NaN     174.21  6037.0  
4                 296425.0     2005.0    6941.39  6037.0  
...                    ...        ...        ...     ...  
2152858           960756.0     2015.0   13494.52  6059.0  
2152859           536061.0     2014.0    6244.16  6059.0  
2152860           208057.0        NaN    5783.88  6059.0  
2152861           424353.0     2015.0    5302.70  6059.0  
2152862           554009.0     2014.0    6761.20  6037.0  

[2152863 rows x 7 columns]

In [13]:
# create acquire function to prepare this nicely
def acquire_zillow():
    """
    Function to pull in zillow data. Returns a DataFrame. Also builds local csv file of the dataset.
    
    Parameters:
    -----------
    none
    
    """
    # build query
    query = """
select 
    bedroomcnt,
    bathroomcnt,
    calculatedfinishedsquarefeet,
    taxvaluedollarcnt,
    yearbuilt,
    taxamount,
    fips
from properties_2017
left join propertylandusetype
    using(propertylandusetypeid)
where propertylandusedesc = 'Single Family Residential'
    """
    
    # use env file to get database url
    url = env.get_db_url('zillow')
    filename = 'zillow.csv'
    
    df = w.check_file_exists(filename,query,url)
    
    return df

In [14]:
# erase the df and start over to check acquire_zillow function
df = []
df

[]

In [15]:
df = acquire_zillow()
w.df_info(df)

Reading from file...


nunique   dtypes  isnull    1774561
bedroomcnt                         19  float64      11       2.00
bathroomcnt                        38  float64      11       2.00
calculatedfinishedsquarefeet    10580  float64    8484    1201.00
taxvaluedollarcnt              592269  float64     493  179757.00
yearbuilt                         153  float64    9337    1964.00
taxamount                      918838  float64    4442    2300.46
fips                                3  float64       0    6059.00

In [16]:
# append the newly built function to write_contents
write_contents.append('''
# create acquire function to prepare this nicely
def acquire_zillow():
    """
    Function to pull in zillow data. Returns a DataFrame. Also builds local csv file of the dataset.
    
    Parameters:
    -----------
    none
    
    """
    # build query
    query = """
select 
    bedroomcnt,
    bathroomcnt,
    calculatedfinishedsquarefeet,
    taxvaluedollarcnt,
    yearbuilt,
    taxamount,
    fips
from properties_2017
left join propertylandusetype
    using(propertylandusetypeid)
where propertylandusedesc = 'Single Family Residential'

    """
    
    # use env file to get database url
    url = env.get_db_url('zillow')
    filename = 'zillow.csv'
    
    df = check_file_exists(filename,query,url)
    
    return df
''')

2. [x] Using your acquired Zillow data, walk through the summarization and cleaning steps in your `wrangle.ipynb` file like we did above. You may handle the missing values however you feel is appropriate and meaningful; remember to document your process and decisions using markdown and code commenting where helpful.

In [18]:
# check out the df using a personal function
w.df_info(df,samples=3)

nunique   dtypes  isnull    1802582     651872  \
bedroomcnt                         19  float64      11       3.00       1.00   
bathroomcnt                        38  float64      11       1.00       1.00   
calculatedfinishedsquarefeet    10580  float64    8484    1235.00     690.00   
taxvaluedollarcnt              592269  float64     493  252208.00  122334.00   
yearbuilt                         153  float64    9337    1920.00    1923.00   
taxamount                      918838  float64    4442    3497.76    1574.04   
fips                                3  float64       0    6037.00    6037.00   

                                 573511  
bedroomcnt                         4.00  
bathroomcnt                        3.00  
calculatedfinishedsquarefeet    2366.00  
taxvaluedollarcnt             452191.00  
yearbuilt                       1948.00  
taxamount                       5567.52  
fips                            6037.00

In [19]:
# Use isnull().mean() to get percentage ratios
round(df.isnull().mean() * 100,3).astype(str)+'%'

bedroomcnt                      0.001%
bathroomcnt                     0.001%
calculatedfinishedsquarefeet    0.394%
taxvaluedollarcnt               0.023%
yearbuilt                       0.434%
taxamount                       0.206%
fips                              0.0%
dtype: object

Considering the ridiculously small proportion of nulls, I'm going to define a rule that if the percent is below 0.01% (or 0.0001, if I'm remembering that properly) then drop the nulls of that column

In [21]:
# get the columns where their proportions are small
auto_drop = [col for col in df.columns if df[col].isna().mean() < 0.001]
auto_drop

['bedroomcnt', 'bathroomcnt', 'taxvaluedollarcnt', 'fips']

In [22]:
# drop the nulls of these and re-examine the dataset
df2 = df.dropna(subset=auto_drop)
w.df_info(df2,include=True)

nunique   dtypes  isnull   1334478      count  \
bedroomcnt                         19  float64       0       2.0  2152359.0   
bathroomcnt                        38  float64       0       2.0  2152359.0   
calculatedfinishedsquarefeet    10578  float64    8099    1721.0  2144260.0   
taxvaluedollarcnt              592265  float64       0  322756.0  2152359.0   
yearbuilt                         153  float64    8958    1967.0  2143401.0   
taxamount                      918821  float64    3971    3272.2  2148388.0   
fips                                3  float64       0    6059.0  2152359.0   

                                       mean            std      min  \
bedroomcnt                         3.287807       0.953796     0.00   
bathroomcnt                        2.231119       0.998916     0.00   
calculatedfinishedsquarefeet    1862.870784    1222.098248     1.00   
taxvaluedollarcnt             461896.824010  699677.417877     1.00   
yearbuilt                       1960.950824      22.161389  1801.00   
taxamount                       5634.944023    8178.944104     1.85   
fips                            6048.374508      20.428394  6037.00   

                                    25%         50%        75%          max  
bedroomcnt                         3.00       3.000       4.00        25.00  
bathroomcnt                        2.00       2.000       3.00        32.00  
calculatedfinishedsquarefeet    1257.00    1623.000    2208.00    952576.00  
taxvaluedollarcnt             188171.00  327671.000  534527.00  98428909.00  
yearbuilt                       1949.00    1958.000    1976.00      2016.00  
taxamount                       2535.06    4108.965    6414.36   1337755.86  
fips                            6037.00    6037.000    6059.00      6111.00

From the comparison, the nuniques went from:
- calculatedfinishedsquarefeet: 10580 -> 10578 (2 val diff)
- taxvaluedollarcnt: 592269 -> 592265 (4 val diff)
- taxamount: 918838 -> 918821 (17 val diff)

Also the nulls went from:
- calculatedfinishedsquarefeet: 8484 -> 8099 (385 diff)
- yearbuilt: 9337 -> 8958 (379 diff)
- taxamount: 4442 -> 3971 (471 diff)

Next let's see about handling that yearbuilt column so we can convert it into an int later.

In [25]:
# maybe we can fill with the average?
df2.yearbuilt.mean()

1960.9508239475488

Doesn't seem too solid. Plus there is a lot of data. We should be safe just dropping them

In [27]:
df = df.dropna()

In [28]:
w.df_info(df,include=True)

nunique   dtypes  isnull     621352      count  \
bedroomcnt                         19  float64       0       5.00  2140235.0   
bathroomcnt                        38  float64       0       3.00  2140235.0   
calculatedfinishedsquarefeet    10557  float64       0    2197.00  2140235.0   
taxvaluedollarcnt              588178  float64       0  848100.00  2140235.0   
yearbuilt                         153  float64       0    1988.00  2140235.0   
taxamount                      915868  float64       0   10124.25  2140235.0   
fips                                3  float64       0    6037.00  2140235.0   

                                       mean            std      min  \
bedroomcnt                         3.301419       0.932600     0.00   
bathroomcnt                        2.240352       0.990549     0.00   
calculatedfinishedsquarefeet    1863.193970    1221.754161     1.00   
taxvaluedollarcnt             460641.625164  677157.635675    22.00   
yearbuilt                       1960.967545      22.150563  1801.00   
taxamount                       5616.711322    7814.562798     6.34   
fips                            6048.309556      20.344910  6037.00   

                                    25%        50%        75%          max  
bedroomcnt                         3.00       3.00       4.00        25.00  
bathroomcnt                        2.00       2.00       3.00        32.00  
calculatedfinishedsquarefeet    1258.00    1623.00    2208.00    952576.00  
taxvaluedollarcnt             189166.00  328296.00  534606.00  90188462.00  
yearbuilt                       1949.00    1958.00    1976.00      2016.00  
taxamount                       2540.85    4111.47    6411.93   1078101.87  
fips                            6037.00    6037.00    6059.00      6111.00

In [29]:
# we'll adjust everything to integer later
# For now let's see if the dollar amounts would be okay to convert to int as well
df.taxamount

4           6941.39
6          10244.94
7           7924.68
11            91.60
14           255.17
             ...   
2152856    13877.56
2152858    13494.52
2152859     6244.16
2152861     5302.70
2152862     6761.20
Name: taxamount, Length: 2140235, dtype: float64

In [30]:
# do a complicated thing to verify that we're safe to convert to integer for this column
# converts the series into string, splits each item into a list,
# gets the second item, and then gets all the value counts
df.taxvaluedollarcnt.astype(str).str.split('.').str[1].value_counts()

taxvaluedollarcnt
0    2140235
Name: count, dtype: int64

In [31]:
# automate the above process for the rest of the columns
for col in df.columns:
    print(df[col].astype(str).str.split('.').str[1].value_counts())
    print()

bedroomcnt
0    2140235
Name: count, dtype: int64

bathroomcnt
0     1910201
5      230031
75          3
Name: count, dtype: int64

calculatedfinishedsquarefeet
0    2140235
Name: count, dtype: int64

taxvaluedollarcnt
0    2140235
Name: count, dtype: int64

yearbuilt
0    2140235
Name: count, dtype: int64

taxamount
24    28892
4     28877
56    28804
7     28795
26    28795
1     28794
8     28745
34    28738
74    28724
14    28703
02    28702
62    28681
94    28650
12    28645
36    28631
08    28630
98    28626
52    28624
92    28617
48    28616
0     28615
82    28604
76    28604
96    28556
44    28540
84    28540
9     28539
72    28531
46    28515
16    28498
28    28494
58    28475
18    28464
04    28444
64    28440
22    28433
2     28428
88    28415
32    28403
78    28389
66    28377
68    28372
86    28355
3     28328
54    28312
38    28306
6     28297
06    28292
5     28258
42    28240
47    14535
51    14491
45    14485
57    14483
83    14468
11    14410
33    143

From the for loop, apparently taxamount and bathroomcnt have decimals, so they're not safe to convert to int

In [33]:
# get a list of the cols
cols = list(df.columns)
cols.remove('taxamount')
cols.remove('bathroomcnt')
cols

['bedroomcnt',
 'calculatedfinishedsquarefeet',
 'taxvaluedollarcnt',
 'yearbuilt',
 'fips']

In [34]:
# convert columns of this list to integer
for col in cols:
    df[col] = df[col].astype(int)

In [35]:
w.df_info(df)

nunique   dtypes  isnull     550607
bedroomcnt                         19    int64       0       2.00
bathroomcnt                        38  float64       0       1.00
calculatedfinishedsquarefeet    10557    int64       0    1071.00
taxvaluedollarcnt              588178    int64       0  165177.00
yearbuilt                         153    int64       0    1946.00
taxamount                      915868  float64       0    1829.73
fips                                3    int64       0    6037.00

In [36]:
df.fips.value_counts()

fips
6037    1425207
6059     552057
6111     162971
Name: count, dtype: int64

In [38]:
# remap fips to counties
df.fips = df.fips.replace({6037:'Los Angeles',6059:'Orange',6111:'Ventura'})

In [39]:
# rename columns with silly naming
df = df.rename(columns={'bedroomcnt':'bedrooms',
                   'bathroomcnt':'bathrooms',
                   'calculatedfinishedsquarefeet':'calculated_area',
                   'fips':'county',
                   'taxvaluedollarcnt':'salesamount'})

In [40]:
def prepare_zillow(df):
    """
    Cleans the zillow DataFrame
    
    Parameters:
    -----------
    df: DataFrame
        - The zillow DataFrame acquired from Codeup's MySQL database (or from local csv)
    
    """
    # re-assign as a copy to fix warnings
    df = df.copy()
    
    # drop the nulls
    df = df.dropna()
    
    # get a list of the cols and drop the two with decimals
    cols = list(df.columns)
    cols.remove('taxamount')
    cols.remove('bathroomcnt')
    
    # convert columns of this list to integer
    for col in cols:
        df[col] = df[col].astype(int)
    
    # convert fips to counties
    df.fips = df.fips.replace({6037:'Los Angeles',6059:'Orange',6111:'Ventura'})

    # rename columns
    df = df.rename(columns={'bedroomcnt':'bedrooms',
                            'bathroomcnt':'bathrooms',
                            'calculatedfinishedsquarefeet':'calculated_area',
                            'fips':'county',
                            'taxvaluedollarcnt':'salesamount'})
    
    return df

In [41]:
# test out 2 functions
df = acquire_zillow()

Reading from file...


In [42]:
df = prepare_zillow(df)

In [43]:
w.df_info(df)

nunique   dtypes  isnull      586
bedrooms              19    int64       0        4
bathrooms             38  float64       0      3.0
calculated_area    10557    int64       0     2460
salesamount       588178    int64       0   448618
yearbuilt            153    int64       0     1978
taxamount         915868  float64       0  5248.44
county                 3   object       0  Ventura

In [44]:
write_contents.append('''
def prepare_zillow(df):
    """
    Cleans the zillow DataFrame
    
    Parameters:
    -----------
    df: DataFrame
        - The zillow DataFrame acquired from Codeup's MySQL database (or from local csv)
    
    """
    # re-assign as a copy to fix warnings
    df = df.copy()
    
    # drop the nulls
    df = df.dropna()
    
    # get a list of the cols and drop the two with decimals
    cols = list(df.columns)
    cols.remove('taxamount')
    cols.remove('bathroomcnt')
    
    # convert columns of this list to integer
    for col in cols:
        df[col] = df[col].astype(int)
    
    # convert fips to counties
    df.fips = df.fips.replace({6037:'Los Angeles',6059:'Orange',6111:'Ventura'})

    # rename columns
    df = df.rename(columns={'bedroomcnt':'bedrooms',
                            'bathroomcnt':'bathrooms',
                            'calculatedfinishedsquarefeet':'calculated_area',
                            'fips':'county',
                            'taxvaluedollarcnt':'salesamount'})
    
    return df
''')

3. [x] Write a function to split your data into train, validate, and test.

In [46]:
from sklearn.model_selection import train_test_split

In [47]:
# importing function previously made
def split_df(df,strat_var,seed=123):
    """
    Returns three dataframes split from one for use in model training, validation, and testing. Takes two arguments:
        df: any dataframe to be split
        strat_var: the value to stratify on. This value should be a categorical variable.
    
    Function performs two splits, first to primarily make the training set, and the second to make the validate and test sets.
    """
    # Run first split
    train, validate_test = train_test_split(df,
                 train_size=0.60,
                random_state=seed,
                 stratify=df[strat_var]
                )
    
    # Run second split
    validate, test = train_test_split(validate_test,
                test_size=0.50,
                 random_state=seed,
                 stratify=validate_test[strat_var]
                )
    
    return train, validate, test

In [48]:
# this yields an error
# train,validate,test = split_df(df,'taxvaluedollarcnt')

In [49]:
train, validate_test = train_test_split(
    df,
    train_size = 0.6,
    random_state = 123
)

In [50]:
train.shape

(1284141, 7)

In [51]:
validate_test.shape

(856094, 7)

In [52]:
validate, test = train_test_split(
    df,
    train_size = 0.5,
    random_state = 123
)

In [2]:
def split_continuous(df,seed=123,**kwargs):
    """
    Returns three dataframes split from one for use in model training, validation, and testing. 
    
    Function performs two splits, first to primarily make the training set, and the second to make the validate and test sets.
    
    Parameters:
    -----------
    df: DataFrame
        - the prepared dataset to be split
    seed: int, defaut=123
        - optional, a seed value to maintain consistency
    """
    parameters={
        'random_state':123,
        'split_1':0.6,
        'split_2':
    }
    # run first split
    train, validate_test = train_test_split(
        df,
        train_size = 0.6,
        random_state = 123
    )
    
    # run second split
    validate, test = train_test_split(
        df,
        train_size = 0.5,
        random_state = 123
    )
    
    return train,validate,test

In [54]:
train,validate,test = split_continuous(df)

In [55]:
train.head()

bedrooms  bathrooms  calculated_area  salesamount  yearbuilt  \
1843306         3        1.0             1598        46546       1954   
1527543         3        3.0             2365       100841       1962   
2057072         2        1.0              929       175454       1946   
1220684         2        1.0             1134       310692       1923   
1886830         3        2.0             1599       616000       1955   

         taxamount       county  
1843306     678.73  Los Angeles  
1527543    1387.90       Orange  
2057072    2268.45  Los Angeles  
1220684    3828.07  Los Angeles  
1886830    7142.49  Los Angeles

In [56]:
write_contents.append('''
def split_continuous(df,seed=123,**kwargs):
    """
    Returns three dataframes split from one for use in model training, validation, and testing. 
    
    Function performs two splits, first to primarily make the training set, and the second to make the validate and test sets.
    
    Parameters:
    -----------
    df: DataFrame
        - the prepared dataset to be split
    seed: int, defaut=123
        - optional, a seed value to maintain consistency
    """
    # run first split
    train, validate_test = train_test_split(
        df,
        train_size = 0.6,
        random_state = 123
    )
    
    # run second split
    validate, test = train_test_split(
        df,
        train_size = 0.5,
        random_state = 123
    )
    
    return train,validate,test
''')

4. [x] Store all of the necessary functions to automate your process from acquiring the data to returning a cleaned dataframe with no missing values in your `wrangle.py` file. Name your final function `wrangle_zillow`.

In [58]:
def wrangle_zillow():
    """
    Acquires, prepares, and splits the zillow dataset.
    
    Parameters:
    -----------
    none
    
    
    """
    
    # acquire the data
    zillow = acquire_zillow()
    
    # prepare the data
    zillow = prepare_zillow(zillow)
    
    # split the data
    train,validate,test = split_continuous(zillow)
    
    return train,validate,test

In [59]:
# wipe df clean
df = []
df

[]

In [60]:
train = []
train

[]

In [61]:
tvt_set = train,validate,test = wrangle_zillow()

Reading from file...


In [62]:
w.df_info(train)

nunique   dtypes  isnull      1403079
bedrooms              17    int64       0            2
bathrooms             37  float64       0          1.0
calculated_area     9210    int64       0          916
salesamount       401250    int64       0        46856
yearbuilt            147    int64       0         1923
taxamount         637157  float64       0       741.17
county                 3   object       0  Los Angeles

In [63]:
write_contents.append('''
def wrangle_zillow():
    """
    Acquires, prepares, and splits the zillow dataset.
    
    Parameters:
    -----------
    none
    
    
    """
    
    # acquire the data
    zillow = acquire_zillow()
    
    # prepare the data
    zillow = prepare_zillow(zillow)
    
    # split the data
    train,validate,test = split_continuous(zillow)
    
    return train,validate,test
''')

In [64]:
# save all the functions made to wrangle.py
with open('wrangle.py','a') as file:
    for content in write_contents:
        file.write(content + '\n')